hack/generator/pkg/codegen/pipeline_determine_resource_ownership.go

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+const resourcesPropertyName = astmodel.PropertyName("Resources")
+
+func determineResourceOwnership() PipelineStage {
+	return PipelineStage{
+		"Determine ARM resource relationships",
+		func(ctx context.Context, definitions astmodel.Types) (astmodel.Types, error) {
+			return determineOwnership(definitions)
+		},
+	}
+}
+
+func determineOwnership(definitions astmodel.Types) (astmodel.Types, error) {
+
+	updatedDefs := make(astmodel.Types)
+
+	for _, def := range definitions {
+		if resourceType, ok := def.Type().(*astmodel.ResourceType); ok {
+			specDef, err := getResourceSpecDefinition(definitions, def.Name(), resourceType)
+			if err != nil {
+				return nil, err
```

Comment:
Add the definition as context to these errors? Otherwise errors might be difficult to pin down to a specific definition. Rather than adding `errors.Wrap(err, def.Name())` to all of them you could extract the body of the loop into another func.

Reply:
Actually, further down it seems like you put lots of identifying information into the errors from the leaf functions so maybe this isn't necessary. Although it can be nice in general - can simplify the error reporting further down (since you don't need to pass contextual info down for the errors) and means that you're guaranteed to get the contextual info, even if the error comes from somewhere that didn't put all of the detail needed to track it down.

Reply:
I went through things and found a few places where there was a parameter that was only used in the error logging, and when I found that I pulled it out and added the context in the caller instead.